src/server/transaction.cc

```diff
@@ -31,6 +31,47 @@ atomic_uint64_t op_seq{1};
 
 constexpr size_t kTransSize [[maybe_unused]] = sizeof(Transaction);
 
+void AnalyzeTxQueue(const EngineShard* shard, const TxQueue* txq) {
+  unsigned q_limit = absl::GetFlag(FLAGS_tx_queue_warning_len);
+  if (txq->size() > q_limit) {
+    static thread_local time_t last_log_time = 0;
+    // TODO: glog provides LOG_EVERY_T, which uses precise clock.
+    // We should introduce inside helio LOG_PERIOD_ATLEAST macro that takes seconds and
+    // uses low precision clock.
+    time_t now = time(nullptr);
+    if (now >= last_log_time + 10) {
+      last_log_time = now;
+      EngineShard::TxQueueInfo info = shard->AnalyzeTxQueue();
+      string msg =
+          StrCat("TxQueue is too long. Tx count:", info.tx_total, ", armed:", info.tx_armed,
+                 ", runnable:", info.tx_runnable, ", total locks: ", info.total_locks,
+                 ", contended locks: ", info.contended_locks, "\n");
+      absl::StrAppend(&msg, "max contention score: ", info.max_contention_score,
+                      ", lock: ", info.max_contention_lock_name,
+                      ", poll_executions:", shard->stats().poll_execution_total);
+      const Transaction* cont_tx = shard->GetContTx();
+      if (cont_tx) {
+        absl::StrAppend(&msg, " continuation_tx: ", cont_tx->DebugId(), " ",
+                        cont_tx->IsArmedInShard(shard->shard_id()) ? " armed" : "");
+      }
+
+      LOG(WARNING) << msg;
+    }
+  }
+}
+
+void RecordTxScheduleStats(const Transaction* tx) {
+  auto* ss = ServerState::tlocal();
```

Comment:
please keep `DCHECK(ss)`. I remember having a bug in one of the unit tests and I needed to run gdb to find the place.